From ebf3bb6ae3a559e6fc7d7fe613cb49aa8e1080f7 Mon Sep 17 00:00:00 2001
From: zirain <hejianpeng2@huawei.com>
Date: Fri, 29 Apr 2022 21:29:46 +0800
Subject: istioctl analyze should add kube sources in all revisions (#38659)

* istioctl analyze should add kube sources in all revisions

* add release-notes
---
 pkg/config/analysis/local/istiod_analyze.go | 24 ++++++++++++++++++++-
 releasenotes/notes/38659.yaml               |  8 +++++++
 2 files changed, 31 insertions(+), 1 deletion(-)
 create mode 100644 releasenotes/notes/38659.yaml

diff --git a/pkg/config/analysis/local/istiod_analyze.go b/pkg/config/analysis/local/istiod_analyze.go
index 7144aaaa87..f238cc5d1a 100644
--- a/pkg/config/analysis/local/istiod_analyze.go
+++ b/pkg/config/analysis/local/istiod_analyze.go
@@ -30,6 +30,7 @@
 
 	"istio.io/api/annotation"
 	"istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/istioctl/pkg/tag"
 	"istio.io/istio/pilot/pkg/config/aggregate"
 	"istio.io/istio/pilot/pkg/config/file"
 	"istio.io/istio/pilot/pkg/config/kube/crdclient"
@@ -261,7 +262,28 @@ func (sa *IstiodAnalyzer) AddReaderKubeSource(readers []ReaderSource) error {
 // AddRunningKubeSource adds a source based on a running k8s cluster to the current IstiodAnalyzer
 // Also tries to get mesh config from the running cluster, if it can
 func (sa *IstiodAnalyzer) AddRunningKubeSource(c kubelib.Client) {
-	sa.AddRunningKubeSourceWithRevision(c, "default")
+	for _, tag := range listRevision(c).UnsortedList() {
+		sa.AddRunningKubeSourceWithRevision(c, tag)
+	}
+}
+
+func listRevision(c kubelib.Client) sets.Set {
+	revisions := sets.New("default")
+	tagWebhooks, err := tag.GetTagWebhooks(context.Background(), c)
+	if err != nil {
+		return revisions
+	}
+
+	for _, wh := range tagWebhooks {
+		tagName, err := tag.GetWebhookTagName(wh)
+		if err != nil {
+			continue
+		}
+
+		revisions.Insert(tagName)
+	}
+
+	return revisions
 }
 
 func (sa *IstiodAnalyzer) AddRunningKubeSourceWithRevision(c kubelib.Client, revision string) {
diff --git a/releasenotes/notes/38659.yaml b/releasenotes/notes/38659.yaml
new file mode 100644
index 0000000000..a4186c6dab
--- /dev/null
+++ b/releasenotes/notes/38659.yaml
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: installation
+issue:
+  - 38148
+releaseNotes:
+  - |
+    **Fixed** istioctl should add kubernetes resource in all revisions when running analyze.
-- 
2.35.3

